render/mapping.go

```diff
@@ -49,52 +33,132 @@ type MapFunc func(report.NodeMetadata) (RenderableNode, bool)
+// RenderableNodes.
+//
+// If this function is given a pseudo node, then it will just return it;
+// Pseudo nodes will never have pids in them, and therefore will never
+// be able to be turned into a Process node.
+//
+// Otherwise, this function will produce a node with the correct ID
+// format for a process, but without any Major or Minor labels.
+// It does not have enough info to do that, and the resulting graph
+// must be merged with a process graph to get that info.
+func MapEndpoint2Process(n RenderableNode) (RenderableNode, bool) {
+	if n.Pseudo {
+		return n, true
+	}
+
+	pid, ok := n.NodeMetadata["pid"]
+	if !ok {
+		// TODO: Propogate a pseudo node instead of dropping this?
+		return RenderableNode{}, false
 	}
 
-	return newRenderableNode(id, major, minor, rank), true
+	id := fmt.Sprintf("pid:%s:%s", getHostname(n.NodeMetadata), pid)
+	return newDerivedNode(id, n), true
 }
 
-// MapContainerIdentity maps container topology node to container mapped nodes.
-func MapContainerIdentity(m report.NodeMetadata) (RenderableNode, bool) {
-	var id, major, minor, rank string
-	if m["docker_container_id"] == "" {
-		id, major, minor, rank = "uncontained", "Uncontained", "", "uncontained"
-	} else {
-		id, major, minor, rank = m["docker_container_id"], m["docker_container_name"], m["domain"], m["docker_image_id"]
+// MapProcess2Container maps process RenderableNodes to container
+// RenderableNodes.
+//
+// If this function is given a node without a docker_container_id
+// (including other pseudo nodes), it will produce an "Uncontained"
+// pseudo node.
+//
+// Otherwise, this function will produce a node with the correct ID
+// format for a container, but without any Major or Minor labels.
+// It does not have enough info to do that, and the resulting graph
+// must be merged with a container graph to get that info.
+func MapProcess2Container(n RenderableNode) (RenderableNode, bool) {
+	id, ok := n.NodeMetadata["docker_container_id"]
+	if !ok || n.Pseudo {
+		return newDerivedPseudoNode(UncontainedID, UncontainedMajor, n), true
 	}
 
-	return newRenderableNode(id, major, minor, rank), true
+	return newDerivedNode(id, n), true
+}
+
+// MapProcess2Name maps process RenderableNodes to RenderableNodes
+// for each process name.
+//
+// This mapper is unlike the other foo2bar mappers as the intention
+// is not to join the information with another topology.  Therefore
+// it outputs a properly-formed node with labels etc.
+func MapProcess2Name(n RenderableNode) (RenderableNode, bool) {
+	if n.Pseudo {
+		return n, true
+	}
+
+	name, ok := n.NodeMetadata["comm"]
+	if !ok {
+		// TODO: Propogate a pseudo node instead of dropping this?
+		return RenderableNode{}, false
+	}
+
+	node := newDerivedNode(name, n)
+	node.LabelMajor = name
+	node.Rank = name
+	return node, true
+}
+
+func getHostname(m report.NodeMetadata) string {
+	hostname, _, _ := report.ParseNodeID(m[report.HostNodeID])
+	return hostname
 }
```

Comment:
1. If you want to parse a host node ID, better to write report.ParseHostNodeID
2. This function is called getHostname but it's actually getHostID, they're not the same
3. If you want the hostname, you need to extract it from the host node metadata...